test/lietensor/test_lietensor.py

```diff
@@ -0,0 +1,297 @@
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+
+def test_lietensor():
+
+    a = torch.randn(3,3)*0.1
+    x = pp.so3(a)
+    y = pp.randn_so3(3, sigma=0.1, dtype=torch.float64, requires_grad=True, device=device)
+    a = pp.randn_se3(3, sigma=0.1, requires_grad=True)
+    b = pp.SE3_type.randn(3, sigma=0.1, requires_grad=True)
+
+    assert y.is_leaf and x.is_leaf and a.is_leaf and b.is_leaf
+
+    (pp.Log(y.Exp())**2).sin().sum().backward()
+
+    print(y)
+    print(y.lshape)
+    print(y.grad)
+
+    def count_parameters(model):
+        return sum(p.numel() for p in model.parameters() if p.requires_grad)
+
+    class SO3Layer(nn.Module):
+        def __init__(self, n):
+            super().__init__()
+            self.weight = pp.Parameter(pp.randn_so3(n))
+
+        def forward(self, x):
+            return self.weight.Exp() * x
+
+    n = 4
+    epoch = 3
+    net = SO3Layer(n).to(device)
+    print(net.weight)
+    net.weight.data.add_(0.1)
+    print(net.weight)
+    optimizer = torch.optim.SGD(net.parameters(), lr = 0.1, momentum=0.9)
+    scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[10,15], gamma=0.1)
+    print(net.weight)
+    for i in range(epoch):
+        optimizer.zero_grad()
+        inputs = pp.randn_SO3(n, device=device)
+        outputs = net(inputs)
+        loss = outputs.abs().sum()
+        loss.backward()
+        optimizer.step()
+        scheduler.step()
+        print(loss)
+    print(net.weight)
+    print("Parameter:", count_parameters(net))
+
+    SO3I = pp.identity_SO3(1, 3, device=device, dtype=torch.float64)
+    so3I = pp.identity_so3(2, 1, device=device, requires_grad=True)
+
+    SE3I = pp.identity_SE3(3, device=device, dtype=torch.float64)
+    se3I = pp.identity_se3(2)
+
+    print(SO3I)
+    print(so3I)
+    print(SE3I)
+    print(se3I)
+
+    idl1 = pp.identity_like(x)
+    idl2 = pp.identity_like(a, requires_grad=True)
+
+    rdl1 = pp.randn_like(x)
+    rdl2 = pp.randn_like(a, sigma=0.1, requires_grad=True, device=device)
+
+    print(rdl1, '\n', rdl2)
+
+    b.Inv()
+    print(b,'\n', b.Inv())
+    print(x,'\n', x.Inv())
+
+    print(pp.Inv(b))
+
+    a = pp.randn_SE3(1,5)
+    b = pp.randn_SE3(5,1)
+    c = pp.randn_SO3(1,5)
+
+    e = a * b
+    f = a * c
+
+    g = pp.randn_so3(1,5)
+    g * 0.1
+
+    pp.Mul(a, b)
+
+    points = torch.randn(2,5,3)
+
+    pt = a * points
+
+    print(pt.shape)
+
+    a = pp.randn_so3(3)
+    I = pp.identity_SO3(3)
+    t = I.Retr(a)
+
+    I = pp.identity_SE3(3)
+    t = I.Retr(a)
+
+    r = pp.Retr(I, a)
+    p = I.Act(a)
+
+    print(r, p)
+
+    X = pp.randn_SE3(8, requires_grad=True, dtype=torch.double)
+    a = pp.randn_se3(8, dtype=torch.double)
+    b = X.Adj(a)
+    assert (b.Exp() * X - X * a.Exp()).abs().mean() < 1e-7
+    J = X.Jinvp(a)
+    print(J)
+
+    X = pp.randn_SE3(6, requires_grad=True)
+    a = pp.randn_se3(6)
+    b = X.AdjT(a)
+    assert (X * b.Exp() - a.Exp() * X).abs().mean() < 1e-6
```

Comment:
A `test_*` function should not include more than one assertion. Otherwise, subsequent tests will not be run if an earlier check fails.

Reply:
I think right now the guideline is not so clear. Maybe we need some professionals to give us some hints on how to come up with good tests. Or the structure for testing large-scale projects.

I have multiple assertions in all my test functions. However, a single test function only focuses on a group of related features. If one assertion fails, it probably means other related assertions may also fail. And it often happens that later assertions are made on results that depends on the previous, failed assertions. If the previous fails, the latter will probably fail. And I strongly recommend that all test scripts should be able to run by themselves. This allows us to only focus on a failed test script, iterating on it without the burden of running all the tests every time.

Reply:
Agreed we should have a guideline. 

> If the previous fails, the latter will probably fail.

I see. If tests are strongly dependent then this is ok. I was trying to say that ideally, we may want each test to only cover one property of the output or even a small piece of code, hence the word "unit". It helps us pin down the problem faster if we know which lines of code cause the problem.

> And I strongly recommend that all test scripts should be able to run by themselves.

Fortunately, pytest has a [`-k` option](https://docs.pytest.org/en/latest/example/markers.html#using-k-expr-to-select-tests-based-on-their-name) for filtering tests by name.